From 02df963c0831381ea85c311862ad1b0fa6090241 Mon Sep 17 00:00:00 2001
From: Max Filippov <jcmvbkbc@gmail.com>
Date: Sun, 9 Nov 2014 06:24:08 +0300
Subject: [PATCH 08/29] xtensa: move callee-saved register check to
 xtensa_call_save_reg

Don't reserve extra stack space with windowed ABI. Save A0 when
profiling is enabled.

Signed-off-by: Max Filippov <jcmvbkbc@gmail.com>
---
 gcc/config/xtensa/xtensa.c | 21 +++++++++++++++------
 1 file changed, 15 insertions(+), 6 deletions(-)

diff --git a/gcc/config/xtensa/xtensa.c b/gcc/config/xtensa/xtensa.c
index bb03030..19f9ee5 100644
--- a/gcc/config/xtensa/xtensa.c
+++ b/gcc/config/xtensa/xtensa.c
@@ -2586,6 +2586,18 @@ xtensa_output_literal (FILE *file, rtx x, enum machine_mode mode, int labelno)
     }
 }
 
+static bool
+xtensa_call_save_reg(int regno)
+{
+  if (TARGET_WINDOWED_ABI)
+    return false;
+
+  if (regno == A0_REG)
+    return crtl->profile || df_regs_ever_live_p (regno);
+
+  return !fixed_regs[regno] && !call_used_regs[regno] &&
+    df_regs_ever_live_p (regno);
+}
 
 /* Return the bytes needed to compute the frame pointer from the current
    stack pointer.  */
@@ -2605,8 +2617,7 @@ compute_frame_size (int size)
   xtensa_callee_save_size = 0;
   for (regno = 0; regno < FIRST_PSEUDO_REGISTER; ++regno)
     {
-      if ((regno == A0_REG || (!fixed_regs[regno] && !call_used_regs[regno])) &&
-	  df_regs_ever_live_p (regno))
+      if (xtensa_call_save_reg(regno))
 	xtensa_callee_save_size += UNITS_PER_WORD;
     }
 
@@ -2692,8 +2703,7 @@ xtensa_expand_prologue (void)
 
   for (regno = 0; regno < FIRST_PSEUDO_REGISTER; ++regno)
     {
-      if ((regno == A0_REG || (!fixed_regs[regno] && !call_used_regs[regno])) &&
-	  df_regs_ever_live_p (regno))
+      if (xtensa_call_save_reg(regno))
         {
 	  rtx x = gen_rtx_PLUS (Pmode, stack_pointer_rtx, GEN_INT (offset));
 
@@ -2780,8 +2790,7 @@ xtensa_expand_epilogue (void)
 
   for (regno = 0; regno < FIRST_PSEUDO_REGISTER; ++regno)
     {
-      if ((regno == A0_REG || (!fixed_regs[regno] && !call_used_regs[regno])) &&
-	  df_regs_ever_live_p (regno))
+      if (xtensa_call_save_reg(regno))
         {
 	  rtx x = gen_rtx_PLUS (Pmode, stack_pointer_rtx, GEN_INT (offset));
 
-- 
2.1.4

